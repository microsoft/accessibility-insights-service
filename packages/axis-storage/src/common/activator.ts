import { injectable } from 'inversify';

// tslint:disable: no-any no-object-literal-type-assertion prefer-object-spread
export class DynamicObject {
    [key: string]: any;
}

@injectable()
export class Activator {
    public convert<T>(source: any, typeT?: new () => T): T {
        if (typeT === undefined) {
            const target = this.createInstance(DynamicObject);

            return <T>Object.assign(target, source);
        } else {
            const target = this.createInstance(typeT);

            return <T>Object.assign(target, source);
        }
    }

    public createInstance<T>(typeT: new () => T): T {
        return new typeT();
    }
}
