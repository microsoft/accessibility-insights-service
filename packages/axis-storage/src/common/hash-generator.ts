import { injectable } from 'inversify';
import * as sha256 from 'sha.js';

@injectable()
export class HashGenerator {
    public constructor(private readonly sha: typeof sha256 = sha256) {}

    public generateBase64Hash(...values: string[]): string {
        const hashSeed: string = values.join('|').toLowerCase();

        return this.sha('sha256')
            .update(hashSeed)
            .digest('hex');
    }

    public getPageScanResultDocumentId(baseUrl: string, url: string, runTimeValue: number): string {
        // preserve parameters order for the hash compatibility
        return this.generateBase64Hash(baseUrl, url, runTimeValue.toString());
    }

    public getScanResultDocumentId(scanUrl: string, selector: string, html: string, resultId: string): string {
        // preserve parameters order for the hash compatibility
        return this.generateBase64Hash(scanUrl, selector, html, resultId);
    }

    public getWebsiteDocumentId(baseUrl: string): string {
        // preserve parameters order for the hash compatibility
        return this.generateBase64Hash(baseUrl);
    }

    public getWebsitePageDocumentId(baseUrl: string, url: string): string {
        // preserve parameters order for the hash compatibility
        return this.generateBase64Hash(baseUrl, url);
    }
}
