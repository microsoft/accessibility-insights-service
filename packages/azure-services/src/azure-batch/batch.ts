// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
/* eslint-disable import/no-internal-modules */
import * as crypto from 'crypto';
import { BatchServiceModels } from '@azure/batch';
import { CloudJob, CloudTask, JobListOptions, OutputFile, TaskListOptions } from '@azure/batch/esm/models';
import { System } from 'common';
import { inject, injectable, optional } from 'inversify';
import * as _ from 'lodash';
import { GlobalLogger } from 'logger';
import { VError } from 'verror';
import { StorageContainerSASUrlProvider } from '../azure-blob/storage-container-sas-url-provider';
import { Message } from '../azure-queue/message';
import { BatchServiceClientProvider, iocTypeNames } from '../ioc-types';
import { BatchConfig } from './batch-config';
import { BatchTaskConfigGenerator } from './batch-task-config-generator';
import { BatchTask, BatchTaskErrorCategory, BatchTaskFailureInfo, getTaskCorrelationId, JobTask, JobTaskState } from './job-task';
import { PoolLoad, PoolMetricsInfo } from './pool-load-generator';

@injectable()
export class Batch {
    public static readonly batchLogContainerName = 'batch-logs';

    public constructor(
        @inject(iocTypeNames.BatchServiceClientProvider) private readonly batchClientProvider: BatchServiceClientProvider,
        @optional()
        @inject(BatchTaskConfigGenerator)
        private readonly batchTaskConfigGenerator: BatchTaskConfigGenerator,
        @inject(StorageContainerSASUrlProvider) private readonly containerSASUrlProvider: StorageContainerSASUrlProvider,
        @inject(BatchConfig) private readonly config: BatchConfig,
        @inject(GlobalLogger) private readonly logger: GlobalLogger,
        // Azure Batch supports the maximum 100 tasks to be added in a single addTaskCollection() API call
        private readonly maxTasks = 100,
    ) {}

    public async getSucceededTasks(jobId: string): Promise<BatchTask[]> {
        const batchTasks: BatchTask[] = [];
        const tasks = await this.getSucceededTaskList(jobId);
        tasks.map((task) => {
            const taskArguments =
                task.environmentSettings !== undefined ? task.environmentSettings.find((e) => e.name === 'TASK_ARGUMENTS') : undefined;

            batchTasks.push({
                id: task.id,
                correlationId: getTaskCorrelationId(task.id),
                taskArguments: taskArguments !== undefined ? taskArguments.value : undefined,
                exitCode: task.executionInfo.exitCode,
                result: task.executionInfo.result,
                timestamp: task.stateTransitionTime,
            });
        });

        return batchTasks;
    }

    public async getFailedTasks(jobId: string): Promise<BatchTask[]> {
        const batchTasks: BatchTask[] = [];
        const tasks = await this.getFailedTaskList(jobId);
        tasks.map((task) => {
            const taskArguments =
                task.environmentSettings !== undefined ? task.environmentSettings.find((e) => e.name === 'TASK_ARGUMENTS') : undefined;

            let failureInfo: BatchTaskFailureInfo;
            if (task.executionInfo.failureInfo !== undefined) {
                let message = '';
                if (task.executionInfo.failureInfo.details !== undefined) {
                    task.executionInfo.failureInfo.details.forEach((details) => {
                        message = `${message}${details.name}: ${details.value}, `;
                    });
                    message = message.slice(0, -2);
                } else {
                    message = task.executionInfo.failureInfo.message;
                }

                failureInfo = {
                    category: task.executionInfo.failureInfo.category as BatchTaskErrorCategory,
                    code: task.executionInfo.failureInfo.code,
                    message,
                };
            }

            batchTasks.push({
                id: task.id,
                correlationId: getTaskCorrelationId(task.id),
                taskArguments: taskArguments !== undefined ? taskArguments.value : undefined,
                exitCode: task.executionInfo.exitCode,
                result: task.executionInfo.result,
                failureInfo,
                timestamp: task.stateTransitionTime,
            });
        });

        return batchTasks;
    }

    public async getPoolMetricsInfo(): Promise<PoolMetricsInfo> {
        const maxTasksPerPool = await this.getMaxTasksPerPool();
        const currentPoolLoad = await this.getCurrentPoolLoad();

        return {
            id: this.config.poolId,
            maxTasksPerPool: maxTasksPerPool,
            load: currentPoolLoad,
        };
    }

    public async createJobIfNotExists(jobId: string, addJobIdIndexOnCreate: boolean = false): Promise<string> {
        let serviceJobId = jobId;
        const client = await this.batchClientProvider();

        try {
            const cloudJob = await client.job.get(serviceJobId);
            if (cloudJob.state !== 'active') {
                throw new VError(`The job ${serviceJobId} is not active and cannot be used to run new tasks.`);
            }
        } catch (error) {
            if ((<BatchServiceModels.BatchError>(<unknown>error)).code === 'JobNotFound') {
                if (addJobIdIndexOnCreate) {
                    serviceJobId = `${jobId}_${crypto.randomBytes(5).toString('hex')}`;
                }

                const jobAddParameter: BatchServiceModels.JobAddParameter = {
                    id: serviceJobId,
                    poolInfo: {
                        poolId: this.config.poolId,
                    },
                    onAllTasksComplete: 'terminatejob',
                };

                await client.job.add(jobAddParameter);

                this.logger.logInfo(`New batch job created successfully.`, { batchJobId: serviceJobId });
            } else {
                throw new VError(error as Error, `An error occurred while retrieving state of ${jobId} job.`);
            }
        }

        return serviceJobId;
    }

    public async createTasks(jobId: string, queueMessages: Message[]): Promise<JobTask[]> {
        const tasks: JobTask[] = [];

        const chunks = System.chunkArray(queueMessages, this.maxTasks);
        await Promise.all(
            chunks.map(async (chunk) => {
                const taskCollection = await this.addTaskCollection(jobId, chunk);
                tasks.push(...taskCollection);
            }),
        );

        return tasks;
    }

    private async getMaxTasksPerPool(): Promise<number> {
        const client = await this.batchClientProvider();
        const poolInfo = await client.pool.get(this.config.poolId);

        return poolInfo.taskSlotsPerNode * (poolInfo.currentDedicatedNodes + poolInfo.currentLowPriorityNodes);
    }

    private async getCurrentPoolLoad(): Promise<PoolLoad> {
        const activeJobIds = await this.getActiveJobIds();

        let activeTasks = 0;
        let runningTasks = 0;
        await Promise.all(
            activeJobIds.map(async (jobId) => {
                const pendingTasks = await this.getPendingTaskList(jobId);
                const pendingTaskInfoList = pendingTasks.map((task) => {
                    return { id: task.id, state: task.state };
                });

                activeTasks += pendingTaskInfoList.filter((task) => task.state === 'preparing' || task.state === 'active').length;
                runningTasks += pendingTaskInfoList.filter((task) => task.state === 'running').length;
            }),
        );

        return {
            activeTasks: activeTasks,
            runningTasks: runningTasks,
        };
    }

    private async getPendingTaskList(jobId: string): Promise<CloudTask[]> {
        const filterClause = `state ne 'completed'`;

        return this.getTaskList(jobId, { filter: filterClause });
    }

    private async getSucceededTaskList(jobId: string): Promise<CloudTask[]> {
        const filterClause = `state eq 'completed' and executionInfo/result eq 'success'`;

        return this.getTaskList(jobId, { filter: filterClause });
    }

    private async getFailedTaskList(jobId: string): Promise<CloudTask[]> {
        const filterClause = `state eq 'completed' and executionInfo/result eq 'failure'`;

        return this.getTaskList(jobId, { filter: filterClause });
    }

    private async getTaskList(jobId: string, options?: TaskListOptions): Promise<CloudTask[]> {
        const tasks = [];
        const taskOptions = {
            taskListOptions: options,
        };

        const client = await this.batchClientProvider();
        const taskListResponse = await client.task.list(jobId, taskOptions);
        tasks.push(...taskListResponse.values());

        let odatanextLink = taskListResponse.odatanextLink;
        while (odatanextLink !== undefined) {
            const taskListResponseNext = await client.task.listNext(odatanextLink, taskOptions);
            tasks.push(...taskListResponseNext.values());
            odatanextLink = taskListResponseNext.odatanextLink;
        }

        return tasks;
    }

    private async getActiveJobIds(): Promise<string[]> {
        const filterClause = `state eq 'active' and executionInfo/poolId eq '${this.config.poolId}'`;
        const jobs = await this.getJobList({ filter: filterClause });

        return jobs.map((i) => i.id);
    }

    private async getJobList(options?: JobListOptions): Promise<CloudJob[]> {
        const jobs = [];
        const listOptions = {
            jobListOptions: options,
        };

        const client = await this.batchClientProvider();
        const jobListResponse = await client.job.list(listOptions);
        jobs.push(...jobListResponse.values());

        let odatanextLink = jobListResponse.odatanextLink;
        while (odatanextLink !== undefined) {
            const jobListResponseNext = await client.job.listNext(odatanextLink, listOptions);
            jobs.push(...jobListResponseNext.values());
            odatanextLink = jobListResponseNext.odatanextLink;
        }

        return jobs;
    }

    private async addTaskCollection(jobId: string, messages: Message[]): Promise<JobTask[]> {
        if (messages.length === 0) {
            return [];
        }

        const jobTasks: Map<string, JobTask> = new Map();
        const taskAddParameters: BatchServiceModels.TaskAddParameter[] = [];
        let sasUrl: string;

        try {
            sasUrl = await this.containerSASUrlProvider.generateSASUrl(Batch.batchLogContainerName);
        } catch (error) {
            this.logger.logError(`Encountered the error while generating Blob Storage SAS URL`, {
                error: System.serializeError(error),
                blobContainerName: Batch.batchLogContainerName,
            });
        }

        await Promise.all(
            messages.map(async (message) => {
                const jobTask = new JobTask(message.messageId);
                jobTasks.set(jobTask.id, jobTask);
                const taskAddParameter = await this.getTaskAddParameter(jobId, jobTask.id, message.messageText, sasUrl);
                taskAddParameters.push(taskAddParameter);
            }),
        );

        const client = await this.batchClientProvider();
        const taskAddCollectionResult = await client.task.addCollection(jobId, taskAddParameters);
        taskAddCollectionResult.value.forEach((taskAddResult) => {
            if (/success/i.test(taskAddResult.status)) {
                jobTasks.get(taskAddResult.taskId).state = JobTaskState.queued;
            } else {
                jobTasks.get(taskAddResult.taskId).state = JobTaskState.failed;
                jobTasks.get(taskAddResult.taskId).error = taskAddResult.error.message.value;
                this.logger.logError(`An error occurred while adding new task to the job.`, {
                    correlatedBatchJobId: jobId,
                    correlatedBatchTaskId: taskAddResult.taskId,
                    error: taskAddResult.error.message.value,
                    taskAddResult: JSON.stringify(taskAddResult),
                });
            }
        });

        return Array.from(jobTasks.values());
    }

    private async getTaskAddParameter(
        jobId: string,
        taskId: string,
        messageText: string,
        sasUrl: string,
    ): Promise<BatchServiceModels.TaskAddParameter> {
        const taskParameter = await this.batchTaskConfigGenerator.getTaskConfigWithImageSupport(
            this.config.accountName,
            taskId,
            messageText,
        );
        if (taskParameter === undefined) {
            return taskParameter;
        }

        if (!_.isNil(sasUrl)) {
            taskParameter.outputFiles = this.getOutFilesConfiguration(jobId, taskId, sasUrl);
        }

        return taskParameter;
    }

    private getOutFilesConfiguration(jobId: string, taskId: string, sasUrl: string): OutputFile[] {
        return [
            {
                filePattern: `../std*.txt`,
                destination: {
                    container: {
                        path: `${jobId}/${taskId}`,
                        containerUrl: sasUrl,
                    },
                },
                uploadOptions: { uploadCondition: 'taskcompletion' },
            },
        ];
    }
}
