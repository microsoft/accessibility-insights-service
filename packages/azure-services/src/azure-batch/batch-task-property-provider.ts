// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BatchServiceModels } from '@azure/batch';
import { ServiceConfiguration, TaskRuntimeConfig } from 'common';
import { inject, injectable, unmanaged } from 'inversify';
import { cloneDeep } from 'lodash';

@injectable()
export abstract class BatchTaskPropertyProvider {
    private readonly taskArgumentsValueName = 'TASK_ARGUMENTS';

    constructor(
        @inject(ServiceConfiguration)
        protected readonly serviceConfig: ServiceConfiguration,
        @unmanaged()
        protected readonly scriptToInvoke: string,
        @unmanaged()
        protected readonly commandLineParameters: string[],
    ) {}

    public getCommandLine(taskArgsString: string): string {
        const taskArgs = JSON.parse(taskArgsString) as { [key: string]: string };

        return this.getCommandLineTemplate().replace(/%(\w*)%/g, (match, key: string) =>
            taskArgs.hasOwnProperty(key) ? taskArgs[key] : '',
        );
    }

    public getAllEnvironmentSettings(taskArgs: string): BatchServiceModels.EnvironmentSetting[] {
        const taskArguments = { name: this.taskArgumentsValueName, value: taskArgs };
        const environmentSettingsCopy = cloneDeep(this.getEnvironmentSettings());
        environmentSettingsCopy.push(taskArguments);

        return environmentSettingsCopy;
    }

    public getResourceFiles(): BatchServiceModels.ResourceFile[] {
        const resourceFiles: BatchServiceModels.ResourceFile[] = [];

        const containerNames = this.getResourceContainerNames();

        containerNames.forEach(containerName => {
            resourceFiles.push({
                autoStorageContainerName: containerName,
            });
        });

        return resourceFiles;
    }

    public async getDefaultTaskConfig(): Promise<TaskRuntimeConfig> {
        return this.serviceConfig.getConfigValue('taskConfig');
    }

    protected abstract getEnvironmentSettings(): BatchServiceModels.EnvironmentSetting[];

    protected abstract getResourceContainerNames(): string[];

    private getCommandLineTemplate(): string {
        let argsPlaceHolders = '';

        this.commandLineParameters.forEach(p => {
            argsPlaceHolders += ` "%${p}%"`;
        });

        return `/bin/bash -c \'${this.scriptToInvoke}${argsPlaceHolders}\'`;
    }
}
