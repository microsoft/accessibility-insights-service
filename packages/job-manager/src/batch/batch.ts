// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BatchServiceModels } from '@azure/batch';
import { Message } from 'azure-services';
import * as crypto from 'crypto';
import { inject, injectable } from 'inversify';
import * as _ from 'lodash';
import { Logger } from 'logger';
import * as moment from 'moment';
import { VError } from 'verror';
import { BatchServiceClientProvider, jobManagerIocTypeNames } from '../job-manager-ioc-types';
import { BatchConfig } from './batch-config';
import { JobTask, JobTaskState } from './job-task';
import { RunnerTaskConfig } from './runner-task-config';

@injectable()
export class Batch {
    public static readonly MAX_TASK_DURATION: number = 10;
    private readonly jobTasks: Map<string, JobTask> = new Map();

    public constructor(
        @inject(BatchConfig) private readonly config: BatchConfig,
        @inject(RunnerTaskConfig) private readonly runnerTaskConfig: RunnerTaskConfig,
        @inject(jobManagerIocTypeNames.BatchServiceClientProvider) private readonly batchClientProvider: BatchServiceClientProvider,
        @inject(Logger) private readonly logger: Logger,
    ) {}

    public async createJobIfNotExists(jobId: string, addJobIdIndexOnCreate: boolean = false): Promise<string> {
        let serviceJobId = jobId;
        const client = await this.batchClientProvider();

        await client.job
            .get(serviceJobId)
            .then(cloudJob => {
                if (cloudJob.state !== 'active') {
                    throw new VError(`The job ${serviceJobId} is not active and cannot be use to run new tasks.`);
                }
            })
            .catch(async (error: BatchServiceModels.BatchError) => {
                if (error.code === 'JobNotFound') {
                    if (addJobIdIndexOnCreate) {
                        serviceJobId = `${jobId}_${crypto.randomBytes(5).toString('hex')}`;
                    }

                    const jobAddParameter: BatchServiceModels.JobAddParameter = {
                        id: serviceJobId,
                        poolInfo: {
                            poolId: this.config.poolId,
                        },
                        onAllTasksComplete: 'terminatejob',
                    };

                    await client.job.add(jobAddParameter);
                    this.logger.logInfo(`New job ${serviceJobId} created.`);
                } else {
                    throw new VError(error as Error, `An error occurred while retrieving state of ${jobId} job.`);
                }
            });

        return serviceJobId;
    }

    public async createTasks(jobId: string, messages: Message[]): Promise<JobTask[]> {
        const taskAddParameters: BatchServiceModels.TaskAddParameter[] = [];

        messages.forEach(message => {
            const jobTask = new JobTask(message.messageId);
            this.jobTasks.set(jobTask.id, jobTask);
            const taskAddParameter = this.getTaskAddParameter(jobTask.id, message.messageText);
            taskAddParameters.push(taskAddParameter);
        });

        if (taskAddParameters.length > 0) {
            const client = await this.batchClientProvider();
            const taskAddCollectionResult = await client.task.addCollection(jobId, taskAddParameters);
            taskAddCollectionResult.value.forEach(taskAddResult => {
                if (/success/i.test(taskAddResult.status)) {
                    this.jobTasks.get(taskAddResult.taskId).state = JobTaskState.queued;
                    this.logger.logInfo(`New task ${taskAddResult.taskId} added to the job ${jobId}.`);
                } else {
                    this.jobTasks.get(taskAddResult.taskId).state = JobTaskState.failed;
                    this.jobTasks.get(taskAddResult.taskId).error = taskAddResult.error.message.value;
                    this.logger.logError(`An error occurred while adding new task ${JSON.stringify(taskAddResult)} to the job ${jobId}.`);
                }
            });
        } else {
            this.logger.logInfo(`No new tasks added to the job ${jobId}.`);
        }

        return Array.from(this.jobTasks.values());
    }

    public async waitJob(jobId: string, pullIntervalMilliseconds: number = 10000): Promise<void> {
        this.logger.logInfo(`Waiting for job ${jobId} to complete.`);
        const client = await this.batchClientProvider();

        return new Promise(async (resolve, reject) => {
            const taskListOptions = {
                filter: `state ne '${JobTaskState.completed}'`,
            };
            const timerId = setInterval(async () => {
                await client.task
                    .list(jobId, { taskListOptions: taskListOptions })
                    .then(async (result: BatchServiceModels.CloudTaskListResult) => {
                        if (result.length === 0 || (result.length === 1 && result[0].id === process.env.AZ_BATCH_TASK_ID)) {
                            clearInterval(timerId);
                            this.logger.logInfo(`Job ${jobId} completed.`);
                            resolve();
                        } else {
                            this.logger.logInfo(`Job ${jobId} in progress with ${result.length} pending tasks.`);
                        }
                    })
                    .catch((error: Error) => {
                        clearInterval(timerId);
                        reject(new VError(error, `An error occurred while retrieving the task list for the job ${jobId}`));
                    });
                // tslint:disable-next-line: align
            }, pullIntervalMilliseconds);
        });
    }

    public async getCreatedTasksState(jobId: string): Promise<JobTask[]> {
        const client = await this.batchClientProvider();
        const cloudTaskListResult = await client.task.list(jobId);
        this.setTasksState(cloudTaskListResult);
        let nextLink = cloudTaskListResult.odatanextLink;
        while (!_.isNil(nextLink)) {
            nextLink = await this.getTasksStateNext(nextLink);
        }

        return Array.from(this.jobTasks.values());
    }

    private async getTasksStateNext(nextPageLink: string): Promise<string> {
        if (!_.isNil(nextPageLink)) {
            const client = await this.batchClientProvider();
            const cloudTaskListResult = await client.task.listNext(nextPageLink);
            this.setTasksState(cloudTaskListResult);

            return cloudTaskListResult.odatanextLink;
        }

        return undefined;
    }

    private setTasksState(cloudTaskList: BatchServiceModels.CloudTaskListResult): void {
        cloudTaskList.forEach(task => {
            if (this.jobTasks.has(task.id)) {
                this.jobTasks.get(task.id).state = task.state;
                this.jobTasks.get(task.id).result = task.executionInfo.result;
                this.logger.logInfo(`Task ${task.id} completed with ${task.executionInfo.result}`);
            }
        });
    }

    private getTaskAddParameter(jobTaskId: string, messageText: string): BatchServiceModels.TaskAddParameter {
        const message = JSON.parse(messageText);
        const commandLine = this.runnerTaskConfig.getCommandLine(message);

        return {
            id: jobTaskId,
            commandLine: commandLine,
            resourceFiles: this.runnerTaskConfig.getResourceFiles(),
            environmentSettings: this.runnerTaskConfig.getEnvironmentSettings(),
            constraints: { maxWallClockTime: moment.duration({ minute: Batch.MAX_TASK_DURATION }).toISOString() },
        };
    }
}
