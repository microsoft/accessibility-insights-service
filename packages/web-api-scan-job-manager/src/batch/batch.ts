// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BatchServiceModels } from '@azure/batch';
// tslint:disable-next-line: no-submodule-imports
import { OutputFile } from '@azure/batch/esm/models';
import { Message, StorageContainerSASUrlProvider } from 'azure-services';
import { ServiceConfiguration, System, TaskRuntimeConfig } from 'common';
import * as crypto from 'crypto';
import { inject, injectable } from 'inversify';
import * as _ from 'lodash';
import { Logger } from 'logger';
import * as moment from 'moment';
import { VError } from 'verror';
import { BatchServiceClientProvider, webApiJobManagerIocTypeNames } from '../web-api-job-manager-ioc-types';
import { BatchConfig } from './batch-config';
import { JobTask, JobTaskState } from './job-task';
import { PoolLoad, PoolMetricsInfo } from './pool-load-generator';
import { RunnerTaskConfig } from './runner-task-config';

@injectable()
export class Batch {
    public static readonly batchLogContainerName = 'batch-logs';

    public constructor(
        @inject(ServiceConfiguration) private readonly serviceConfig: ServiceConfiguration,
        @inject(BatchConfig) private readonly config: BatchConfig,
        @inject(RunnerTaskConfig) private readonly runnerTaskConfig: RunnerTaskConfig,
        @inject(webApiJobManagerIocTypeNames.BatchServiceClientProvider) private readonly batchClientProvider: BatchServiceClientProvider,
        @inject(Logger) private readonly logger: Logger,
        @inject(StorageContainerSASUrlProvider) private readonly containerSASUrlProvider: StorageContainerSASUrlProvider,
    ) {}

    public async getPoolMetricsInfo(): Promise<PoolMetricsInfo> {
        const maxTasksPerPool = await this.getMaxTasksPerPool();
        const currentPoolLoad = await this.getCurrentPoolLoad();

        return {
            id: this.config.poolId,
            maxTasksPerPool: maxTasksPerPool,
            load: currentPoolLoad,
        };
    }

    public async createJobIfNotExists(jobId: string, addJobIdIndexOnCreate: boolean = false): Promise<string> {
        let serviceJobId = jobId;
        const client = await this.batchClientProvider();

        try {
            const cloudJob = await client.job.get(serviceJobId);
            if (cloudJob.state !== 'active') {
                throw new VError(`The job ${serviceJobId} is not active and cannot be used to run new tasks.`);
            }
        } catch (error) {
            if ((<BatchServiceModels.BatchError>(<unknown>error)).code === 'JobNotFound') {
                if (addJobIdIndexOnCreate) {
                    serviceJobId = `${jobId}_${crypto.randomBytes(5).toString('hex')}`;
                }

                const jobAddParameter: BatchServiceModels.JobAddParameter = {
                    id: serviceJobId,
                    poolInfo: {
                        poolId: this.config.poolId,
                    },
                    onAllTasksComplete: 'terminatejob',
                };

                await client.job.add(jobAddParameter);

                this.logger.logInfo(`New job ${serviceJobId} created.`);
            } else {
                throw new VError(error as Error, `An error occurred while retrieving state of ${jobId} job.`);
            }
        }

        return serviceJobId;
    }

    public async createTasks(jobId: string, messages: Message[]): Promise<JobTask[]> {
        const tasks: JobTask[] = [];

        // Azure Batch supports the maximum 100 tasks to be added in addTaskCollection() API call
        const chunks = System.chunkArray(messages, 100);
        await Promise.all(
            chunks.map(async chunk => {
                const taskCollection = await this.addTaskCollection(jobId, chunk);
                tasks.push(...taskCollection);
            }),
        );

        return tasks;
    }

    private async getMaxTasksPerPool(): Promise<number> {
        const client = await this.batchClientProvider();
        const poolInfo = await client.pool.get(this.config.poolId);

        return poolInfo.maxTasksPerNode * (poolInfo.currentDedicatedNodes + poolInfo.currentLowPriorityNodes);
    }

    private async getCurrentPoolLoad(): Promise<PoolLoad> {
        const activeJobIds = await this.getActiveJobIds();

        let activeTasks = 0;
        let runningTasks = 0;

        const client = await this.batchClientProvider();
        await Promise.all(
            activeJobIds.map(async jobId => {
                const jobInfo = await client.job.getTaskCounts(jobId);
                activeTasks += jobInfo.active;
                runningTasks += jobInfo.running;
            }),
        );

        return {
            activeTasks: activeTasks,
            runningTasks: runningTasks,
        };
    }

    private async getActiveJobIds(): Promise<string[]> {
        const filterClause = `state eq 'active' and executionInfo/poolId eq '${this.config.poolId}'`;
        const options = {
            jobListOptions: { filter: filterClause },
        };

        const jobs = [];
        const client = await this.batchClientProvider();
        const jobListResponse = await client.job.list(options);
        jobs.push(...jobListResponse.values());

        let odatanextLink = jobListResponse.odatanextLink;
        while (odatanextLink !== undefined) {
            const jobListResponseNext = await client.job.listNext(odatanextLink, options);
            jobs.push(...jobListResponseNext.values());
            odatanextLink = jobListResponseNext.odatanextLink;
        }

        return jobs.map(i => i.id);
    }

    private async addTaskCollection(jobId: string, messages: Message[]): Promise<JobTask[]> {
        if (messages.length === 0) {
            return [];
        }

        const jobTasks: Map<string, JobTask> = new Map();
        const taskAddParameters: BatchServiceModels.TaskAddParameter[] = [];
        const maxTaskDurationInMinutes = (await this.getTaskConfig()).taskTimeoutInMinutes;
        let sasUrl: string;

        try {
            sasUrl = await this.containerSASUrlProvider.generateSASUrl(Batch.batchLogContainerName);
        } catch (error) {
            this.logger.logError(`encounter error while generating sas url ${error}`);
        }

        messages.forEach(message => {
            const jobTask = new JobTask(message.messageId);
            jobTasks.set(jobTask.id, jobTask);
            const taskAddParameter = this.getTaskAddParameter(jobId, jobTask.id, message.messageText, maxTaskDurationInMinutes, sasUrl);
            taskAddParameters.push(taskAddParameter);
        });

        const client = await this.batchClientProvider();
        const taskAddCollectionResult = await client.task.addCollection(jobId, taskAddParameters);
        taskAddCollectionResult.value.forEach(taskAddResult => {
            if (/success/i.test(taskAddResult.status)) {
                jobTasks.get(taskAddResult.taskId).state = JobTaskState.queued;
            } else {
                jobTasks.get(taskAddResult.taskId).state = JobTaskState.failed;
                jobTasks.get(taskAddResult.taskId).error = taskAddResult.error.message.value;
                this.logger.logError(`An error occurred while adding new task ${JSON.stringify(taskAddResult)} to the job ${jobId}.`);
            }
        });

        return Array.from(jobTasks.values());
    }

    private getTaskAddParameter(
        jobId: string,
        jobTaskId: string,
        messageText: string,
        maxTaskDurationInMinutes: number,
        sasUrl: string,
    ): BatchServiceModels.TaskAddParameter {
        const message = JSON.parse(messageText);
        const commandLine = this.runnerTaskConfig.getCommandLine(message);
        const taskParameter: BatchServiceModels.TaskAddParameter = {
            id: jobTaskId,
            commandLine: commandLine,
            resourceFiles: this.runnerTaskConfig.getResourceFiles(),
            environmentSettings: this.runnerTaskConfig.getEnvironmentSettings(),
            constraints: { maxWallClockTime: moment.duration({ minute: maxTaskDurationInMinutes }).toISOString() },
        };

        if (!_.isNil(sasUrl)) {
            taskParameter.outputFiles = this.getOutFilesConfiguration(jobId, jobTaskId, sasUrl);
        }

        return taskParameter;
    }

    private getOutFilesConfiguration(jobId: string, jobTaskId: string, sasUrl: string): OutputFile[] {
        return [
            {
                filePattern: `../std*.txt`,
                destination: {
                    container: {
                        path: `${jobId}/${jobTaskId}`,
                        containerUrl: sasUrl,
                    },
                },
                uploadOptions: { uploadCondition: 'taskcompletion' },
            },
        ];
    }

    private async getTaskConfig(): Promise<TaskRuntimeConfig> {
        return this.serviceConfig.getConfigValue('taskConfig');
    }
}
