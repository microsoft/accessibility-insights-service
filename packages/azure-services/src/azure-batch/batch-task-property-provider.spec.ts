// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { BatchServiceModels } from '@azure/batch';
import { ServiceConfiguration } from 'common';
import { IMock, Mock } from 'typemoq';
import { BatchTaskPropertyProvider } from './batch-task-property-provider';

// tslint:disable: no-object-literal-type-assertion no-any

class TestableBatchTaskPropertyProvider extends BatchTaskPropertyProvider {
    constructor(serviceConfig: ServiceConfiguration) {
        super(serviceConfig, 'script1.sh', ['arg1Key', 'arg2Key']);
    }

    public getEnvironmentSettings(): BatchServiceModels.EnvironmentSetting[] {
        return [
            {
                name: 'env1',
                value: 'env1 value',
            },
            {
                name: 'env2',
                value: 'env2 value',
            },
        ];
    }

    protected getResourceContainerNames(): string[] {
        return ['container1', 'container2'];
    }
}

describe(BatchTaskPropertyProvider, () => {
    let testSubject: TestableBatchTaskPropertyProvider;
    let serviceConfig: IMock<ServiceConfiguration>;
    let taskArgs: { [key: string]: string };

    beforeEach(() => {
        serviceConfig = Mock.ofType(ServiceConfiguration);
        testSubject = new TestableBatchTaskPropertyProvider(serviceConfig.object);

        taskArgs = {
            arg1Key: 'arg1 value',
            arg2Key: 'arg2 value',
            someUnknownKey: 'unknown value',
        };
    });

    describe('getCommandLine', () => {
        it('generates command line string', () => {
            expect(testSubject.getCommandLine(JSON.stringify(taskArgs))).toMatchSnapshot();
        });
    });

    describe('getAllEnvironmentSettings', () => {
        it('gets environment settings with task parameters', () => {
            expect(testSubject.getAllEnvironmentSettings(JSON.stringify(taskArgs))).toMatchSnapshot();
        });
    });

    describe('getResourceFiles', () => {
        it('returns resource files', () => {
            expect(testSubject.getResourceFiles()).toMatchSnapshot();
        });
    });

    describe('getTaskConfig', () => {
        it('returns task config', async () => {
            const taskConfig = 'task config' as any;
            serviceConfig.setup(async s => s.getConfigValue('taskConfig')).returns(async () => Promise.resolve(taskConfig));

            await expect(testSubject.getDefaultTaskConfig()).resolves.toBe(taskConfig);
        });
    });
});
