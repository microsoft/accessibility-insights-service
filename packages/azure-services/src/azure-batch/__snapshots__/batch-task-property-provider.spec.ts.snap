// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchTaskPropertyProvider getAllEnvironmentSettings gets environment settings with task parameters 1`] = `
Array [
  Object {
    "name": "env1",
    "value": "env1 value",
  },
  Object {
    "name": "env2",
    "value": "env2 value",
  },
  Object {
    "name": "TASK_ARGUMENTS",
    "value": "{\\"arg1Key\\":\\"arg1 value\\",\\"arg2Key\\":\\"arg2 value\\",\\"someUnknownKey\\":\\"unknown value\\"}",
  },
]
`;

exports[`BatchTaskPropertyProvider getCommandLine generates command line string 1`] = `"/bin/bash -c 'script1.sh \\"arg1 value\\" \\"arg2 value\\"'"`;

exports[`BatchTaskPropertyProvider getResourceFiles returns resource files 1`] = `
Array [
  Object {
    "autoStorageContainerName": "container1",
  },
  Object {
    "autoStorageContainerName": "container2",
  },
]
`;
